### DISCUSSION TITLE
Unbricking Flashforge Dreamer

### DISCUSSION THREAD

**Discussion by freeifier** [moonglow/FlashForge_Marlin#120] [Q&A]
Hey I wanted to configure the firmware differently changed two parameters. After I built and flashed it via fw update trigger it is stuck in the boot screen, so I guess I messed something up. I just ordered an ST Link adapter, because I tried firmware flash tool on linux and also flashprint but I was not able to get into bootloader mode as it seems.

I read through some articles and but I don't  know what I woudl have to do when my adapter arrives.
Maybe you could give me some help and probably others can use this as reference for unbricking their machines.

Cheers

**Comment by moonglow** [comment]
Hi @freeifier !

Please provide information about your printer model first.  Some printers does not support update via bootloader ( as example Dremel 3D20 ) and if you update it to your own custom firmware it can soft break your printer... it is not a problem.. but need some time to order st-link or something like that to bring your printer back to life.

**Comment by freeifier** [comment]
So St link is on the way will be here tomorrow I guess.

I have a Flashforhe dreamer for sure. Also had Marlin up and running. i also reflashed it via Firmware update trigger to different Marlin Version with dir swap both worked.

But when trying to flash my custom Firmware.
 write file success was displayed but printer now is stuck in Boot screen. 

I want to reflash Software with St link.

**Reply by moonglow** [reply]
Hm... looks like your Dreamer has an old bootloader without update support... If our script updater under Linux does not works ( it does not give you any information ? ). Anyway if you will flash it, i recommends you also to update printer bootloader ( it supports update via bootloader mode, you will able refresh your firmware from any state without jtag ). You can get new bootloader image here:  #116

**Reply by freeifier** [reply]
Ah yes that would be awesome. Maybe you could hell me through that Seen you gave some explanation to another guy.

**Reply by moonglow** [reply]
yeh no problem

**Comment by dr-mav** [comment]
Hello,

I have FF dreamer. After the first attempt to install the Marlin firmware I brick the printer. The process of upgrade went successfully, but printer never restarted into the new firmware. Is there a way to unbrick the printer? 

PS I have very limited skills in programming so please use short words.

**Reply by moonglow** [reply]
Hi @dr-mav !

Can you provide us step by step what you did before ?

@freeifier 

Yeah you can do it already, first of all you can backup your current bootloader how it explain [HERE](https://github.com/moonglow/FlashForge_Marlin/wiki/Backup-printer-firmware). I also recommends you to install [STM32CubeProg](https://www.st.com/en/development-tools/stm32cubeprog.html) it is GUI for ST-Link very easy and intuitive to use.

**Reply by dr-mav** [reply]
Hello @moonglow 
I downloaded your repository. Installed VS. After that fixed the .h file. Build the .bin file. And using your instruction and fwriter flashed the printer. Got the message on the screen "write file successful" and nothing after that. Waited maybe 20 minutes, restarted and nothing pass the blue screen with logo.

**Comment by freeifier** [comment]
@moonglow So my ST link is here, I guess i would connect it up. I also have openocd installed and then I will update bootloader and try flashing working marlin version ? 

@dr-mav I guess when I figured it out I could help you to outline the approach you would have to take the bottom panel of and then buy a ST link adapter looks like a usb stick and connect it up to the mainboard (4 wires). Using a programm on your computer you then can refllash the software to unbrick your printer.

**Reply by dr-mav** [reply]
@freeifier thank you. the panel is off and waiting. Let me know when you will have an outlined process.

**Reply by freeifier** [reply]
I used somthing like this https://www.ebay.de/itm/255283221952?hash=item3b7010d5c0:g:Uq0AAOSwYRJdQWxj this seems to be a knock off but worked fine for me, you will have to connect it like here https://github.com/moonglow/FlashForge_Marlin/wiki/Backup-printer-firmware
1.
then I guess st link cube programmer might be the easiest software to do so, for me I flashed bootloader and then the firmware something like here https://github.com/moonglow/FlashForge_Marlin/discussions/101. 
It first did not seem to boot into marlin it just got me error as in the 101 discussion and it always flashed the fw from the built in sd card. 
2.
In the discussion they said to delete the .bin file on the sd card and so then I put the "official" build from moonglow renamed it dreamer.bin put it where to old file was under sys/dreamer.bin onto the sdcard, rebooted the printer and it flashed the printer then flashed the fw.
3.
Then I put the sd card back into the pc deleted the dreamer.bin file put it back into the printer and now it boots. Its not perfect the bootloader seems to have an issue with the fw checking but via this workaround my printer is already back up and printing right now.

**Reply by freeifier** [reply]
@dr-mav If some steps are unclear hit me up

**Reply by dr-mav** [reply]
Hello @freeifier I got the same ST-link as yours. Do I need connect VCC? How to be sure that my bootloader is gone? Right now if I will unplug TF card, I'm going to see black screen with text "Bootloader v1.2.201400822. Copyright 2014 by Flashforge. TF card Fail."

**Reply by freeifier** [reply]
You will need to connect vcc to  3,3V from the St link adapter otherwise the CPU of the printer will have no power.

I guess @moonglow knows more about the bootloader but in my understanding you sure can use your existing bootloader it should not be damaged by any flashing you did beforehand, but if you continue using it you will have to encrypt firmware, it is not that hard there is a guide here for that too. I flashd the different bootloader because I wanted to eliminate encryption makes things easier. But I also created a backup of the original bootloader as described in the backup post.

**Reply by dr-mav** [reply]
Thank you @freeifier for your help. Did you solder the pin to VCC pad on SD card reader, or just remove the WiFi module and plug the pin into the 01 - WIFI_VCC_33?

**Reply by freeifier** [reply]
soldered it to vcc from sd

**Reply by dr-mav** [reply]
Will do tonight.

**Reply by dr-mav** [reply]
By some reason I'm not able to get Cube Programmer to connect to the motherboard. Do you know the size of that 3pin connector on the motherboard?
What I do wrong?
![Screenshot 2022-09-13 225643](https://user-images.githubusercontent.com/113116768/190048346-9e679148-04b6-44b9-aca3-41cc2099645b.png)

**Reply by freeifier** [reply]
Did you try connecting with the jumper wires which were included, these are really bad for that connector type, I broke of the plastic piece of the jumpers on one side and then insulated the metal connectors revealed from that with some heat shrink. After that I bent them a bit smaller so they fit nice and tight over the pins. Otherwise it might also be a good idea to get the right connector yes, but I do not know what size it is.

**Reply by freeifier** [reply]
And I recommend you do a backuo even tho there is something broken maybe you wan't to swtich back to original bootloader at some point.

**Reply by dr-mav** [reply]
Thank you for quick reply. I thought that the connection is maybe a problem. Will try to investigate that. I will let you know if that is a problem.

**Reply by dr-mav** [reply]
Looks like it is not a connection problem. What else can be an issue?

**Reply by tracedgod** [reply]
> Looks like it is not a connection problem. What else can be an issue?

How are you powering the board? I used the 24v from the dreamer's PSU to get it to fully turn on and had no issues. Are you getting any lights on the coreboard itself when it's plugged in?

**Reply by dr-mav** [reply]
@tracedgod you connected just SWDio and SWclk with main printer power turned on, didn't you?

**Reply by dr-mav** [reply]
@tracedgod when I conect 3.3V, ground and SWdio with SWclk I have 2 green LED and screen is on.

**Reply by dr-mav** [reply]
@freeifier @moonglow Ok, looks like I have more luck with old STM32 ST-link software. I think I have done something wrong again. No the screen is black and no messages on it. Can someone explains me as 2years old what exactly I have to do? I tryied to save the .bin of the original firmware, but it is just 4Kb.

**Reply by freeifier** [reply]
> @tracedgod you connected just SWDio and SWclk with main printer power turned on, didn't you?

It is important that you connect the ground of the stlink Programmer with the ground on the board, otherwise the stlink can't make a connection to the processor

**Reply by dr-mav** [reply]
@freeifier Thank you. I Think I made it. Ok here is what I did, maybe it will help someone:
1) Get St-Link adapter (https://www.amazon.com/Aideepen-ST-Link-Programming-Emulator-Downloader/dp/B01J7N3RE6/ref=sr_1_5?keywords=st+link+v2&qid=1663451915&sprefix=st+link%2Caps%2C80&sr=8-5)
2) Install STM32 St-link software https://www.st.com/en/development-tools/stsw-link004.html#get-software
3) Make the connection as here https://github.com/moonglow/FlashForge_Marlin/wiki/Backup-printer-firmware Best way for me was solder the VCC 3.3V and the Ground wire to the SD card PCB. Take the other 2 wires of the original connector from your adaptor and remove the black plastic. Use the shrink tube or tape to isolate the female connector (maybe pinch a little bit for tight connection). Use them to connect to 1&2 pin of the debugging port (SWClK and SWDIO).
![image](https://user-images.githubusercontent.com/113116768/190878643-f968af24-9be2-49bf-8509-bdb51d0287bc.png)
4) Connect your adaptor (do not connect it to the printer yet) and update its firmware through the STM32 St-Link if needed. 
5) Disconnect adapter from your PC.
6) Remove the TF card
7) Now connect adaptor to the printer and then to the PC.
8) Open settings (check pictures) and set them as on my screenshots. My issue was that I wasn't able to flash under normal mode.
![image](https://user-images.githubusercontent.com/113116768/190878662-26ff793e-1dab-4f04-b5c7-377147e4ac0f.png)
![image](https://user-images.githubusercontent.com/113116768/190878693-142db834-2325-4e4f-b022-e49dca864ba2.png)
![image](https://user-images.githubusercontent.com/113116768/190878706-45e94848-4700-4f49-840f-76f2bca88b78.png)

9) Now click Ctrl+P "program & verify" 
![image](https://user-images.githubusercontent.com/113116768/190878715-8b07bfda-0098-4683-91a4-0bc1f147546b.png)

10) Choose the [dreamer_bootloader_v1.4_20161121_no_check](https://github.com/moonglow/FlashForge_Marlin/files/8244294/dreamer_bootloader_v1.4_20161121_no_check.zip)
11) Click "Start"
![image](https://user-images.githubusercontent.com/113116768/190878223-a3d03d9f-dcc8-493f-b5ad-04ea87bcce39.png)
12) Wait when the flash process is done it should take about 2 sec. You will be returned to the main screen. The printer's screen should be black but the backlight is on. Remove the adaptor and disconnect it from your printer.
12) Now we need to replace the official FF dreamer.bin with @moonglow firmware (https://github.com/moonglow/FlashForge_Marlin/releases/download/v0.14.3/color_ui_04242022.zip) at the TF card (F:/sys/). Keep the name of the .bin file the same (**dreamer.bin**). 
13) Reinstall the TF card into the printer. 
14) Turn on the main power of the printer.
15) You will see the "check ID fail" and shortly after printer will start flashing process. White bar will fill the bottom of the screen until "write file success" message.  (like here https://github.com/moonglow/FlashForge_Marlin/discussions/98)
16) Restart your printer. The black screen will appear and in less then 10 sec you should see the marlin logo.
17) Done! Printer is ready.
![image](https://user-images.githubusercontent.com/113116768/190878651-b7dee463-8b93-4dc5-bdae-1179568b1441.png)

Big thanks to @freeifier and @moonglow (Спасибо большое дружище! Жду дальнейших обновлений)

**Reply by freeifier** [reply]
Glas to see it works, good job. Maybe wie should put you steps into a guide its well explained.
But did you leave the .bin on the SD card ? Because then mine would always try to flash it and never Boot.

**Reply by tracedgod** [reply]
Good to see you got it! I still can't get mine to boot marlin, it will write the file but does not want to boot, not sure whats going on with that. I have submitted an issue for this: #121 

> But did you leave the .bin on the SD card ? Because then mine would always try to flash it and never Boot.
>
Mine does this as well, I've disassembled the boot-loader bin and from what I can tell, it looks like it's always checking for that dreamer.bin, and if it finds it on the SD card, it is going to flash it without any consideration if it has already done so. Doesn't seem like there is any sort of update mode programmed into the firmware, still mapping out everything though so if I find anything that can stop that, I'll post it. I was considering writing a custom boot-loader for the dreamer, not sure if I will/need to though.

Also, I was able to get rid of the error messages on the boot screen, if you want the boot-loader to get rid of those messages you can grab it from #121.

**Reply by dr-mav** [reply]
@tracedgod after the printer flashed itself from the TF card I removed it. In my case the screen stays black for 5-10 sec before marlin logo shows up. No blue FF screen anymore. If I will figure out how to dump the whole firmware from chip I can attach mine here. 
PS what revision your main board?

**Reply by tracedgod** [reply]
@dr-mav i have a RevH CoreBoard. When I flash marlin, it never will get to the black screen portion for whatever reason, just sits on the blue FlashForge screen.

**Reply by dr-mav** [reply]
> @dr-mav i have a RevH CoreBoard. When I flash marlin, it never will get to the black screen portion for whatever reason, just sits on the blue FlashForge screen.

I had issue with my programmer. My first one or didn't connect at all or didn't able to finish the programming. 

@moonglow  maybe we all should share our bootloaders here so people can use them in case the original is not working.

Plus, my usb stopped working. I think it is issue with Win10 and drivers. Dis anyone have that problem?

**Reply by tracedgod** [reply]
> > @dr-mav i have a RevH CoreBoard. When I flash marlin, it never will get to the black screen portion for whatever reason, just sits on the blue FlashForge screen.
> 
> I had issue with my programmer. My first one or didn't connect at all or didn't able to finish the programming.
> 
> @moonglow maybe we all should share our bootloaders here so people can use them in case the original is not working.
> 
> Plus, my usb stopped working. I think it is issue with Win10 and drivers. Dis anyone have that problem?

The USB problem was my main issue. Does your device show up in Device Manager at all when it's plugged in? Mine did not. I have since solved the issue by purchasing a new CoreBoard from eBay. As @moonglow mentioned, the MCU USB block was probably fried. I was able to use the program to write the firmware via USB with the new board, but I did have to flash the no-check bootloader to it first (the new board came with a weird firmware installed, looks like it might have been used internally to test the boards at FlashForge?).

**Reply by dr-mav** [reply]
@tracedgod In my case I can see unknow USB device with error code 43. As I understand it is a problem of drivers in win10. I'm not sure though. 
Additionally maybe some one has nice settings for the dreamer (with 2 nozzle) in Cura. I'm trying to nailed it down, but still not as quite good results as I had with original firmware and Simlify3D.

**Reply by tracedgod** [reply]
> @tracedgod In my case I can see unknow USB device with error code 43. As I understand it is a problem of drivers in win10. I'm not sure though. Additionally maybe some one has nice settings for the dreamer (with 2 nozzle) in Cura. I'm trying to nailed it down, but still not as quite good results as I had with original firmware and Simlify3D.

Gotcha, ~~I think~~ you need to install the [STM32 Virtual COM Port Driver](https://www.st.com/en/development-tools/stsw-stm32102.html), as marlin uses the virtual COM port for USB connection ~~I believe~~. (Edit: I have confirmed that Marlin indeed does show as the STM32 Virtual COM Port device.) As for the profile, I'm still sort of working on it as I go, but I was able to get my 2 extruders to show up and all that. There are some changes I had to make to the Start + End GCodes as well. I will commit the files to a repo when I get a chance.

**Comment by freeifier** [comment]
So I used cube programmer to flash bootloader and flash marlin firmware I have a similar issue as this guy Damaged boot loader on PowerSpec? #98 but it works when deleting the bin on the sd.

**Comment by moonglow** [comment]
@dr-mav 
> "After the first attempt to install the Marlin firmware I brick the printer."
> "I  have very limited skills in programming so please use short words."

But actually you do not use pretested prebuiled firmware and try to:

>"After that fixed the .h file. Build the .bin file."

Thats why i ask you about more info...  I assume you forget to encrypt your firmware before start to flash it to printer...  You can try our Linux firmware upgrade script and if your printer bootloader is not too old you will flash it... another way to unbrick your printer is using St-Link programmer as @freeifier already tells before.

------------------------
@freeifier 
> "Its not perfect the bootloader seems to have an issue with the fw checking but via this workaround"

Hm... whats wrong with that bootloader ?

**Reply by freeifier** [reply]
I always get check error fail and if there is firmware on sd then it will always flash it. So after flashing I deleted the dreamer.bin as mentioned in the post I got thsi from and then it boots fine.

**Reply by moonglow** [reply]
@freeifier !!!! thanx ... i will check it and fix such issue

**Reply by dr-mav** [reply]
Hello @moonglow I probably missed that part about encryption or it was in reedme file.

**Reply by tracedgod** [reply]
> I always get check error fail and if there is firmware on sd then it will always flash it. So after flashing I deleted the dreamer.bin as mentioned in the post I got thsi from and then it boots fine.

Having the same issue. I have not been able to connect to my dreamer via USB to install marlin for a few months now, unsure why. FF support told me it could be firmware related but they didn't know how to fix it.. so I figured I would try my luck with the ST-Link adapter and re-flash the factory firmware first to see if it fixed my USB issue. Unfortunately, even though I took a backup of my original bootloader, it does not boot at all when reflashing. I can get it to boot with the bootloader from #116 ~~but it just sits on the check error fail and does not boot at all, along with some other garbled mess on the screen. I have also tried placing an original, still encrypted firmware on the internal SD card in sys/dreamer.bin, a decrypted firmware, marlin firmware, none of them will boot at all even after I delete the bin file from the internal SD card.~~

EDIT: I just got it to boot back into the original firmware. I used an encrypted dreamer firmware (pulled from flashprint files, v2.15) and left it on the file write success screen for about 10 minutes. I then turned off the printer, and removed the dreamer.bin file on my internal SD and put it back into the printer. I then left it turned off for another 20 minutes, turned it back on, and it was booted up to the factory firmware. Unfortunately, after all that still no USB connection... 😢

I have attached my original firmware dump in case you wanted to take a look at it: [ff-backup.zip](https://github.com/moonglow/FlashForge_Marlin/files/9562189/ff-backup.zip)

**Reply by moonglow** [reply]
@dr-mav 
Yep rule is simple... if you want to load firmware via ff bootloader you must encrypt firmwre first ( my build script ( `marlin_builder.sh` ) will do it automatically, but if you build on your own you must ensure to encrypt it by yourself too ) and if you write firmware directly via JTAG it must be raw ( no encryption needed ).  

@tracedgod 
USB is not firmware related problem ( Marlin or FF nevermind ), sometimes it can be USB socket/soldering problem, so you can check wire connection from USB cable side to MCU pins and sometimes or your MCU USB block was damaged by static electricity or something like that ( some users report me that FF support team change whole motherboard in such cases ).... 

p.s: you ff_backup is only 4KB i think you save only part of firmware that you saw on flasher screen. Anyway make sure what other users can't run your bootloader because it linked to your motherboard thats why i provide patched one...

**Reply by tracedgod** [reply]
> @tracedgod USB is not firmware related problem ( Marlin or FF nevermind ),

I didn't think so either, just wanted to try it out to see if it was potentially an issue. I think there was an issue with my original bootloader anyways, it never would update when I put dreamer.bin on the internal SD before I flashed the patched one. 

> sometimes it can be USB socket/soldering problem, so you can check wire connection from USB cable side to MCU pins and sometimes or your MCU USB block was damaged by static electricity or something like that ( some users report me that FF support team change whole motherboard in such cases )....

Sounds like the USB issue is probably going to be the MCU block as I have already replaced the port in the past with verification that all pins have continuity to their respective CPU pins, ~~I'll look into possibly sourcing a replacement if at all possible (the block doesn't have any sort of labeling on it).~~ EDIT: I read that wrong, you meant the actual peripheral block in the processor itself.

> p.s: you ff_backup is only 4KB i think you save only part of firmware that you saw on flasher screen. Anyway make sure what other users can't run your bootloader because it linked to your motherboard thats why i provide patched one...

I noticed the backup didn't look right, I probably didn't do it correctly. Oh well, it's booted up fine with the patched bootloader you provided so I'm not really that worried about it.

@moonglow , what base address should I be writing the unecrypted firmware to via SWD/JTAG if I wanted to straight flash marlin? I want to run marlin but I can't load it via USB (though I suppose I could encrypt and put it on the internal SD card as well). Thanks for all the directions + info!
